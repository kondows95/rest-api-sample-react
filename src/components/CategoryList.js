import React from 'react'
import PropTypes from 'prop-types'
import { makeStyles } from '@material-ui/core/styles'
import { 
  Box, Container, Paper, Grid, Button,
  Dialog, DialogTitle, DialogContent, DialogActions, TextField, 
} from '@material-ui/core'
import { validateForm } from '../util'

const useStyles = makeStyles(theme => ({
  button: {
    marginRight: 0,
    marginLeft: "auto",
  },
  paper: {
    display: 'flex',
    margin: theme.spacing(2, 1),
    padding: theme.spacing(2, 2),
  },
}))

const CategoryList = ({ categories, saveCategory, deleteCategory }) => {
  const classes = useStyles()
  const [dialogOpen, setDialogOpen] = React.useState(false)
  const [selectedCategory, setSelectedCategory] = React.useState(null)
  const [isDelete, setIsDelete] = React.useState(false)
  const [errors, setErrors] = React.useState({})
  
  const initialCategory = {id: null, name: ""}
  
  const handleChangeValue = fieldName => event => {
    const newCategory = {...selectedCategory}
    newCategory[fieldName] =  event.target.value
    setSelectedCategory(newCategory)
  }
  
  const handleDelete = category => event => {
    setSelectedCategory(category)
    setDialogOpen(true)
    setIsDelete(true)
    setErrors({})
  }
  
  const handleEdit = category => event => {
    setSelectedCategory(category)
    setDialogOpen(true)
    setIsDelete(false)
    setErrors({})
  }

  const handleCloseDialog = () => {
    setDialogOpen(false)
    setSelectedCategory(null)
  }
  
  const validationSetting = {
    isEmpty: ['name'],
  }
  
  const handleSubmit = () => {
    if (selectedCategory) {
      const errs = validateForm(validationSetting, selectedCategory)
      if (errs) {
        setErrors(errs)
      }
      else {
        if (isDelete) {
          deleteCategory(selectedCategory.id)
        }
        else {
          saveCategory(selectedCategory)
        }
        handleCloseDialog()
      }
    }
  }
  
  const deleteDialog = (selectedCategory && isDelete) ? (
    <Dialog 
      open={dialogOpen} 
      onClose={handleCloseDialog} 
      aria-labelledby="category-delete-dialog"
      fullWidth
      maxWidth="xs"
    >
      <DialogTitle id="category-delete-dialog">
        {"Are you sure?"}
      </DialogTitle>
      <DialogContent>
        <Box>Do you really want to delete {selectedCategory.id}: {selectedCategory.name}.</Box>
        <Box fontWeight={600}></Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleCloseDialog} color="primary">
          Cancel
        </Button>
        <Button onClick={handleSubmit} color="primary">
          Delete
        </Button>
      </DialogActions>
    </Dialog>
  ) : null
  
  const saveDialog = (selectedCategory && isDelete === false) ? (
    <Dialog 
      open={dialogOpen} 
      onClose={handleCloseDialog} 
      aria-labelledby="category-save-dialog"
      fullWidth
      maxWidth="xs"
    >
      <DialogTitle id="category-save-dialog" >
        {selectedCategory.id ? "Edit (ID:"+selectedCategory.id+")" : "Create"}
      </DialogTitle>
      <DialogContent>
        <TextField
          variant="outlined"
          autoFocus
          error={errors.name ? true : false}
          id="name"
          label="Name"
          value={selectedCategory.name}
          onChange={handleChangeValue("name")}
          fullWidth
        />
      </DialogContent>
      <DialogActions>
        <Button id = "cancle" onClick={handleCloseDialog} color="primary">
          Cancel
        </Button>
        <Button id = "submit" onClick={handleSubmit} color="primary">
          Submit
        </Button>
      </DialogActions>
    </Dialog>
  ) : null
  
  const paperItems = []
  for (const category of categories) {
    paperItems.push(
      <Grid key={category.id} item xs={12} sm={4} lg={3}>
        <Paper className={classes.paper}>
          <Box display="flex" flexDirection="column" flexGrow={1}>
            <Box fontWeight={600}>{category.id}: {category.name}</Box>
            <Box ml="auto" mr={0} mt={2}>
              <Button color="primary" onClick={handleEdit(category)}>
                Edit
              </Button>
              <Button color="primary" onClick={handleDelete(category)}>
                Delete
              </Button>
            </Box>
          </Box>
        </Paper>
      </Grid>
    )
  }
  
  const paperControl = (
    <Paper className={classes.paper}>
      <Box ml={0} my="auto" fontWeight={600}>
        Categories ({categories.length})
      </Box>
      <Button 
        id="create"
        variant="contained" 
        color="secondary" 
        className={classes.button}
        onClick={handleEdit(initialCategory)}
      >
        Create
      </Button>
    </Paper>
  )
  
  return (
    <React.Fragment>
      <Container maxWidth="lg">
        {paperControl}
        <Grid container>
          {paperItems}
        </Grid>
      </Container>
      {saveDialog}
      {deleteDialog}
    </React.Fragment>
  )
}

const categoryPropTypes = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
})

CategoryList.propTypes = {
  categories: PropTypes.arrayOf(categoryPropTypes.isRequired).isRequired,
}

export default CategoryList
