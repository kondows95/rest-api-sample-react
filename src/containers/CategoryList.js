import { connect } from 'react-redux'
import CategoryList from '../components/CategoryList'
import { saveCategory, deleteCategory, dialogBox } from '../modules/categories'

export default connect(
  (state) => ({
    items: state.items.rows,
    categories: state.categories.rows,
    loading:state.categories.loading,
    closeDialog:state.categories.closeDialog
  }),
  (dispatch) => ({
    saveCategory: (category) =>  dispatch(saveCategory(category)),
    deleteCategory: (id) =>  dispatch(deleteCategory(id)),
    dialogBox:(value) => dispatch(dialogBox(value)),
  })
)(CategoryList)
