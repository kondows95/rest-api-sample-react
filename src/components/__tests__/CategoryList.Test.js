import React from 'react'
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import CategoryList from '../CategoryList';

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.outerHTML = '<body></body>'
  container = null
})

const valueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, "value").set;
describe("testing creating category", () => {
   it('should exists one create button and one name textfield', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} />, container);
    });
    const button = container.querySelectorAll('Button')
    expect(button[0].id).toBe("create")
    act(() => {
      button[0].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const inputArr = document.querySelectorAll('input');
    expect(inputArr).toHaveLength(1);
    expect(inputArr[0].id).toBe("name");
  });

  it('check the textfield handle change and click cancle button', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} />, container);
    });
    const CreateButton = container.querySelectorAll('Button')
    act(() => {
      CreateButton[0].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const inputArr = document.querySelectorAll('input');
    act(() => {
      valueSetter.call(inputArr[0], 123);
      inputArr[0].dispatchEvent(new Event('change', { bubbles: true }));
    });
    expect(inputArr[0].value).toBe("123");
    const buttonArray = document.querySelectorAll('button');
    act(() => {
      buttonArray[5].dispatchEvent(new Event('click', { bubbles: true }))
    })
    act(() => {
      CreateButton[0].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const inputArr1 = document.querySelectorAll('input');
    expect(inputArr1[0].value).toBe("");
  });
  it('click submit button', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    const saveCategory = jest.fn((setSelectedCategory) => {
      categories = [...categories, setSelectedCategory]
    });
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} saveCategory={saveCategory} />, container);
    });
    const CreateButton = container.querySelectorAll('Button')
    act(() => {
      CreateButton[0].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const inputArr = document.querySelectorAll('input');
    act(() => {
      valueSetter.call(inputArr[0], 'abc');
      inputArr[0].dispatchEvent(new Event('change', { bubbles: true }));
    });
    const buttonArray = document.querySelectorAll('button');
    expect(buttonArray[6].id).toBe("submit")
    act(() => {
      buttonArray[6].dispatchEvent(new Event('click', { bubbles: true }))
    })
    const expectedCategoires = [
      { id: 1, name: 'a' },
      { id: 2, name: 'b' },
      { id: null, name: 'abc' }
    ]
    expect(saveCategory).toHaveBeenCalled();
    expect(categories).toStrictEqual(expectedCategoires)
  });
  it('should not be called if textfield is empty', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    const saveCategory = jest.fn((setSelectedCategory) => {
      categories = [...categories, setSelectedCategory]
    });
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} saveCategory={saveCategory} />, container);
    });
    const CreateButton = container.querySelectorAll('Button')
    act(() => {
      CreateButton[0].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const buttonArray = document.querySelectorAll('button');
    act(() => {
      buttonArray[6].dispatchEvent(new Event('click', { bubbles: true }))
    })
    expect(saveCategory).not.toHaveBeenCalled();
  });
})

describe("testing category list", () => {
  it('testing edit button', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} />, container);
    });
    const button = container.querySelectorAll('Button')
    act(() => {
      button[1].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const inputArr = document.querySelectorAll('input');
    expect(inputArr[0].value).toBe("a");
  });

  it('testing delete button', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    const deleteCategory = jest.fn((category_id) => {
      if (category_id === 1) categories = [{ id: 2, name: "b" }];
      else categories = [{ id: 1, name: "a" }];
    });
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} deleteCategory={deleteCategory} />, container);
    });
    const button = container.querySelectorAll('Button');
    act(() => {
      button[2].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const buttonArr = document.querySelectorAll('Button')
    expect(buttonArr[5].id).toBe("delete cancel");
    expect(buttonArr[6].id).toBe("delete");
    act(() => {
      buttonArr[6].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(deleteCategory).toHaveBeenCalled();
    const expectedCategoires=[{ id: 2, name: "b" }];
    expect(categories).toStrictEqual(expectedCategoires)
  });

  it('testing cancle button', () => {
    let categories = [{ id: 1, name: "a" }, { id: 2, name: "b" }];
    const deleteCategory = jest.fn((category_id) => {
      if (category_id === 1) categories = [{ id: 2, name: "b" }];
      else categories = [{ id: 1, name: "a" }];
    });
    act(() => {
      ReactDOM.render(<CategoryList categories={categories} deleteCategory={deleteCategory} />, container);
    });
    const button = container.querySelectorAll('Button');
    act(() => {
      button[2].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    const buttonArr = document.querySelectorAll('Button')
    act(() => {
      buttonArr[5].dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(deleteCategory).not.toHaveBeenCalled();
  });
})