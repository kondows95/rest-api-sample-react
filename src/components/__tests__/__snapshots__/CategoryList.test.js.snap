// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList component matches the snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-2 MuiPaper-rounded"
  >
    <div
      className="MuiBox-root MuiBox-root-38"
    >
      Categories (
      2
      )
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root makeStyles-button-1 MuiButton-contained MuiButton-containedSecondary"
      disabled={false}
      id="create"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Create
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-2 MuiPaper-rounded"
      >
        <div
          className="MuiBox-root MuiBox-root-162"
        >
          <div
            className="MuiBox-root MuiBox-root-163"
          >
            1
            : 
            a
          </div>
          <div
            className="MuiBox-root MuiBox-root-164"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
              disabled={false}
              id="categoryedit"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Edit
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
              disabled={false}
              id="categorydelete"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Delete
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-2 MuiPaper-rounded"
      >
        <div
          className="MuiBox-root MuiBox-root-165"
        >
          <div
            className="MuiBox-root MuiBox-root-166"
          >
            2
            : 
            b
          </div>
          <div
            className="MuiBox-root MuiBox-root-167"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
              disabled={false}
              id="categoryedit"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Edit
              </span>
            </button>
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
              disabled={false}
              id="categorydelete"
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                Delete
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
